encrypted-media-respec.html

```diff
@@ -2101,6 +2101,30 @@
           <p>The following steps are run:</p>
           <ol>
             <li><p>Let the <var>session</var> be the associated <a>MediaKeySession</a> object.</p></li>
+            <li><p>Run the <a def-id="destroy-keys-and-licenses-algorithm"></a> algorithm.</p></li>
+            <li><p>Run the <a def-id="update-key-statuses-algorithm"></a> algorithm on the <var>session</var>, providing an empty sequence.</p></li>
+            <li><p>Run the <a def-id="update-expiration-algorithm"></a> algorithm on the <var>session</var>, providing <code>NaN</code>.</p></li>
+            <li><p>Let <var>promise</var> be the <a def-id="closed"></a> attribute of the <var>session</var>.</p></li>
+            <li><p>Resolve <var>promise</var>.</p></li>
+          </ol>
+        </section>
+        <section id="destroy-keys">
+          <h4>Destroy Keys and Licenses</h4>
+          <p>
+            The Destroy Keys and Licenses algorithm is run when the CDM needs to destroy all the keys and licenses associated with a session that
+            have not been explicitly stored.
+          </p>
+          <p>
+            For sessions of type <a def-id="temporary-session"></a> and <a def-id="persistent-usage-record-session"></a>, this algorithm MUST be
+            executed when the <a>MediaKeySession</a> object becomes inaccessible to the page, if the <a def-id="session-close-algorithm"></a>
+            algorithm has not already been run.
+          </p>
+          <p class="note">
+            This is necessary to ensure that the keys and licenses associated with the session are correctly destroyed when the session ends.
+          </p>
+          <p>The following steps are run:</p>
+          <ol>
+            <li><p>Let the <var>session</var> be the associated <a>MediaKeySession</a> object.</p></li>
```

Comment:
Regarding line 2135 (applies to the existing spec text too):

This algorithm, like Session Close, is an asynchronous callback from the CDM to the user agent to initiate application-visible behavior. It doesn't make sense for the CDM to call back into the user agent only to have it call back to destroy the keys. Destroying the keys is something that should happen around this time, but a) the user agent can't destroy them, which is what the step currently says, and b) this should have been done before calling the user agent.

All these algorithms have "is run when the CDM..." language, though I don't know if that is the only time they are used (see #153). This is the only one that calls back into the CDM.

One solution is to state that the CDM must have destroyed the (non-persistent) key(s) and license(s) before calling either of these two algorithms. That probably changes the name and purpose of this algorithm.